src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Sockets;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    public class SocketReceiver
+    {
+        private readonly Socket _socket;
+        private readonly SocketAsyncEventArgs _eventArgs = new SocketAsyncEventArgs();
+        private readonly SocketAwaitable _awaitable = new SocketAwaitable();
+
+        public SocketReceiver(Socket socket)
+        {
+            _socket = socket;
+            _eventArgs.UserToken = this;
+            _eventArgs.Completed += (_, e) => ((SocketReceiver)e.UserToken).ReceiveCompleted(e);
+        }
+
+        public SocketAwaitable ReceiveAsync(Buffer<byte> buffer)
+        {
+            if (!buffer.TryGetArray(out var segment))
+            {
+                throw new InvalidOperationException();
```

Comment:
Same as below comment. Put a message in here.

Reply:
https://github.com/aspnet/KestrelHttpServer/blob/659fa967a1900653f7a82f02624c7c7995a3b786/src/Kestrel.Core/Internal/Http/PipelineExtensions.cs#L27

Reply:
Even better :)